src/utils/getFenceAndImportDiffsFromGit.ts

```diff
@@ -0,0 +1,249 @@
+            fencePatches.push(patch);
+        } else if (
+            extensions.some(
+                scriptFileExtension =>
+                    oldFile.path().endsWith(scriptFileExtension) ||
+                    newFile.path().endsWith(scriptFileExtension)
+            )
+        ) {
+            sourcePatches.push(patch);
+        } else {
+        }
+    }
+    return [fencePatches, sourcePatches];
+}
+
+export type SourceImportDiff = {
+    addedImports: string[];
+    removedImports: string[];
+};
+
+export type FenceAndImportDiffs = {
+    fenceDiffs: Map<NormalizedPath, FenceDiff>;
+    sourceImportDiffs: Map<NormalizedPath, SourceImportDiff>;
+};
+
+/**
+ * Given a git OID (commit hash) or ref name (refs/heads/master, HEAD~1),
+ * finds the difference in script file imports and fence changes
+ * between the current git index (e.g. staged set of files),
+ * and the specified oid/refname.
+ *
+ * @param compareOidOrRefName an oid or refname
+ * @returns The fence and import diffs
+ */
+export async function getFenceAndImportDiffsFromGit(
+    compareOidOrRefName: string
+): Promise<FenceAndImportDiffs | null> {
+    const repo = await Git.Repository.open(process.cwd());
+    const [index, headCommitTree, compareTree] = await Promise.all([
+        repo.index(),
+        repo.getHeadCommit().then(headCommit => headCommit?.getTree?.()),
+        resolveToCommit(repo, compareOidOrRefName).then(commit => commit.getTree()),
+    ]);
+
+    let repoDiff: Git.Diff;
+    const indexToHead = await Git.Diff.treeToIndex(repo, headCommitTree, null);
+    const indexIsEmpty = indexToHead.numDeltas() === 0;
+
+    // Permit all extensions in the extension set. If we are
+    // overly-permissive here, the script files we detect for
+    // checking should be filtered out while providing source
+    // files.
+    const mostPermissiveExtensionSet = getScriptFileExtensions({
+        includeJson: true,
+        jsx: true,
+        allowJs: true,
+        // This is used as a glob of for *${dotExt}, so .d.ts files
+        // will be included in by the *.ts glob. Likewise for .d.tsx
+        includeDefinitions: false,
+    });
+
+    if (!indexIsEmpty) {
+        repoDiff = await Git.Diff.treeToIndex(repo, compareTree, index, {
+            contextLines: 0,
+            pathspec: mostPermissiveExtensionSet.map(dotExt => '*' + dotExt),
+        });
+    } else {
+        repoDiff = await Git.Diff.treeToTree(repo, compareTree, headCommitTree, {
+            contextLines: 0,
+            pathspec: mostPermissiveExtensionSet.map(dotExt => '*' + dotExt),
+        });
+    }
+    const [fencePatches, sourcePatches] = await getFenceAndSourcePatches(
+        repoDiff,
+        mostPermissiveExtensionSet
+    );
+
+    // TODO: track files across moves (Should just be a fence removal and addition)
+    for (let patch of [...fencePatches, ...sourcePatches]) {
+        if (patch.oldFile().path() && patch.oldFile().path() !== patch.newFile().path()) {

```

Comment:
Won't this trigger on cases where a file is deleted, too?

Reply:
nodegit represents a delete by setting the oldpath and newpath as the same value, but setting the object hash of the new file as 0. I'll leave a comment to that effect here.